Duplicate project import as a warning only

philderbeast opened the pull request:
Depends-on #11773.

I'll squash commits before applying the merge label if this pull request is approved.

Pretty much the same as #9933 but gives a warning instead of an error when duplicate imports that are not cyclical are detected. I did this work in Oct 2024 but didn't raise a pull request for it then. I was reporting the warning during parsing but needed an `IORef` for that. After Matthew Pickering suggesting it, I'm now doing the reporting after parsing.

- Add a new module `cabal-install/src/Distribution/Client/ProjectConfig/Import.hs` and move some stuff there from `cabal-install-solver/src/Distribution/Solver/Types/ProjectConfigPath.hs` that was only ever used in `cabal-install`.
- Shorten the haddock sections in `Distribution.Client.ProjectConfig.Legacy` so that they don't wrap when rendered.

For messages, I added some doctests and ended up with two ways of comparing `ProjectConfigPath`.

```haskell
-- | A comparison that puts projects first, URLs last and sorts the other paths
-- lexically.
compareLexically :: ProjectConfigPath -> ProjectConfigPath -> Ordering

-- | A comparison that puts projects first, URLs last and sorts the other paths
-- by putting longer paths after shorter ones as measured by the number of path
-- segments. If still equal, then sorting is lexical.
compareSegmentally:: ProjectConfigPath -> ProjectConfigPath -> Ordering
```

I changed the project skeleton, retaining the info we have already (because we have to read the file), that the item is a local file or a URI.

```diff
- type ProjectConfigSkeleton = CondTree ConfVar ([ProjectConfigPath], ProjectConfig)
+ type ProjectConfigSkeleton = CondTree ConfVar ([(Maybe URI, ProjectConfigPath)], ProjectConfig)
```

With that information, I'm able to avoid monitoring remote files:

```diff
- monitorFiles $ map monitorFileHashed (projectConfigPathRoot <$> projectSkeletonImports pcs)
+ monitorFiles
+  [ monitorFileHashed (projectConfigPathRoot path)
+  | (Nothing, path) <- projectSkeletonImports pcs
+  ]
```

I added another type to help classify between, root, file import and URI import. I used this to help detect and report duplicates:

```haskell
-- | Isomorphic with 'ProjectConfigPath' but with separate constructors for the
-- root, imported file and imported URI.
data ProjectNode a where
  ProjectRoot :: FilePath -> ProjectNode ProjectFilePath
  ProjectFileImport :: FilePath -> ProjectConfigPath -> ProjectNode FilePath
  ProjectUriImport :: URI -> ProjectConfigPath -> ProjectNode URI
```

I developed this with `ghc-9.14.1` and required some extra type annotations for prior GHC versions to compile, those in `seenImport` and `instance Ord (ProjectNode a)`.

I moved the following types and functions to the `Import` module too. These were either in the Legacy parser or in both parsers.

```haskell
type ProjectConfigSkeleton = CondTree ConfVar ([(Maybe URI, ProjectConfigPath)], ProjectConfig)

projectSkeletonImports :: ProjectConfigSkeleton -> [(Maybe URI, ProjectConfigPath)]
projectSkeletonImports = fst . view traverseCondTreeA

fetchImportConfig
  :: FilePath
  -> HttpTransport
  -> Verbosity
  -> FilePath
  -> ProjectConfigPath
  -> IO (Maybe URI, BS.ByteString)
```

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
  * [x] [Is the change significant?](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#is-my-change-significant) If so, remember to add `significance: significant` in the changelog file.
* [ ] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

ulysses4ever commented on cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs:
```diff
@@ -203,9 +204,12 @@ import System.FilePath (isAbsolute, isPathSeparator, makeValid, splitFileName, (
   , ($+$)
   )
-import qualified Text.PrettyPrint as Disp
+import qualified Text.PrettyPrint as Disp (empty, int, render, text)
```
`text` again? (see above)

philderbeast replied:
Same goes for `render`. Actually only `empty` needs to be used qualified.

ulysses4ever commented on cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs:
```diff
@@ -256,38 +260,66 @@ parseProject
   :: FilePath
   -> HttpTransport
   -> Verbosity
+  -> IORef (NubList (FilePath, ProjectConfigPath))
```
With all the `Dupes`-business above, we end up with a very low-level type in this signature. Any chance the `NubList (FilePath, ProjectConfigPath)` could be factored out too? A type synonym at least, but the type is suspiciously close to `dupesSeenImportsBy` above, so I expect some economy between the two places.

philderbeast replied:
Thanks for pointing that out. I've used a record now for `(FilePath, ProjectConfigPath)` named `ProjectImport`.

ulysses4ever commented on cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs:
```diff
@@ -256,38 +260,66 @@ parseProject
-    -- NOTE: Reverse the warnings so they are in line number order.
-    <&> \case ProjectParseOk ws x -> ProjectParseOk (reverse ws) x; x -> x
+parseProject rootPath cacheDir httpTransport verbosity configToParse = do
+  let (dir, projectFileName) = splitFileName rootPath
```
It'd be preferable to not change layout/identation without a real need, I think.

philderbeast replied:
I strive for minimal diffs. In this situation, I was rebasing something older and it made the merge conflict resolution easier this way.

https://github.com/haskell/cabal/blame/51e1817dec9980d0d204a051fc3c4dc981c3d027/cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs#L259-L266

https://github.com/haskell/cabal/blame/54d364df92b654ce712f0fe277ed709e710d8bf4/cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs#L246-L250

I actually prefer the way it was before the latest commit, with the `= do` on the same line rather than `do` by itself on its own line. This also has less indentation.

If you're OK with the way it is, then I'd like to keep it that way, the way it was, one commit before the last commit.

philderbeast replied:
I'm a little surprised that fourmolu doesn't always normalize.

philderbeast replied:
Or it could have been I used a more recent version of fourmolu at some stage.

ulysses4ever reviewed (approved):
Thanks!

mpickering commented on cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs:
```diff
@@ -259,48 +256,79 @@ parseProject
   :: FilePath
   -> HttpTransport
   -> Verbosity
+  -> IORef (NubList ProjectImport)
```
Have you considered an approach where the duplicate imports are reported after parsing is completed? It seems to conflate two things together by tracking this state during parsing. 

I might have expected a warning to inspect the `projectConfigProvenance` field and issue warnings based on that (perhaps `Set` is not the right datatype for that). Using `IORef`s here doesn't seem ideal to me.

philderbeast replied:
> Have you considered an approach where the duplicate imports are reported after parsing is completed?

I hadn't and I expect that would be better now that we're aiming to report a warning and not an error that interrupts parsing. Thanks for the suggestion.

philderbeast committed:
Warn on duplicate non-cyclical project imports

- Add Y-forking import test
- A test for detecting when the same config is imported via many different paths
- Error on duplicate imports
- Do the filtering in duplicateImportMsg
- Use duplicateImportMsg for cycles too
- Add haddocks to IORef parameter
- Add changelog entry
- Use ordNub instead of nub
- Use NubList
- Share implement of duplicate and cyclical messages
- Update expectation for non-cyclical duplicate import
- Only show a warning
- Add woops project with a time cost
- Use noticeDoc instead of warn
- Render duplicate imports
- Add Ord instance for Dupes, sort on dupesNormLocPath
- Fixups after rebase
- Satisfy hlint
- Remove -XMultiWayIf
- Remove mention of yops from the changelog
- Satisfy fix-whitespace
- Test with a time cost of duplicate imports
- Fewer imports from PrettyPrint qualified as Disp
- Add data ProjectImport replacing tuples
- Combine fields as ProjectImport
- Rename field to dupesImports
- Add haddocks to Dupes fields
- Mark test as flaky
- Any test accessing stackage seems susceptible
- Move unique duplicates to own test
- Use legacy parser for path duplicates test
- Add foo.cabal package so that packages exist
- Satisfy fix-whitespace
- Use local version of lts-21.25
- Remove repo
- Use </> for expected paths
- Note that this change gives a warning.
- Exclude lts-21.25.config from typos
- Detect duplicate imports after parsing
- Merge upstream
- Update duplicate import counts
- Satisfy fourmolu
- Move duplicate detection to calling module
- Add parsing module haddock section header
- Export only reportDuplicateImports
- Test all project parser options
- Add script tests
- Add haddocks to reportDuplicateImportsh
- Use Data.Function ((&)) pipelining
- Inline do blockh
- Satisfy fix-whitespace
- Redo Ord Dupes
- Try to enforce sorting
- Add docs to ProjectImport
- Add instances & doctests for sorting ProjectImport
- Fix sorting of ProjectConfigPath
- Update test expectation with correct sorting
- Compare versions of compare
- Update cabal.out for dedup tests
- Rename to compareForDisplay
- Add more detail to docProjectConfigFiles docs
- Add haddocks and rename comparisons
- Don't repeat the root comment
- Pattern match on NonEmpty not List
- Sort expectation by import chain length
- Use GADTs for ProjectImport
- Satisfy hlint, remove unused LANGUAGE pragmas
- Bubble up Maybe URI
- Add ProjectRoot constructorh
- Rename ProjectImport to ProjectNode
- Add Maybe URI to project skeleton
- Use ProjectFilePath
- Satisfy hlint
- Add projectNodes
- Redo detectDupes
- Move duplicate imports to Import module
- Move fetchImportConfig to Import module
- Don't export parseProjectSkeleton
- Don't re-export ProjectConfigSkeleton
- Remove unnecessary primes
- Add splitImports
- Remove ProjectNodes and projectNodes
- Move stuff to Client.ProjectConfig.Import
- Fixup after rebase
- Satisfy fourmolu
- Don't monitor (Nothing :: Maybe URI)
- Add type annotation for older ghc versions
- Setup for doctests
- Need -XConstraintKinds with older ghcsh
- Update .out files for UniquePathDuplicates
- Don't repeat recordMode
- Remove +legacy-comparison
- Note significance in changelog and movements
- Less indent
- Remove redundant comparison
- Hide newtype ProjectFilePath
- Don't export duplicateImportMsg or ProjectNode
- Gut lts-21.25.config but leave notes

philderbeast commented:
I'm seeing validate failures due to repeated "Warning: ..." messages that I'm not seeing when not using the validate build. I'm able to reproduce those duplicated messages if I run with the `./dist-newstyle-validate-ghc-9.14.1/.../cabal` version.

```
$ ./validate.sh --run-cli-suite
...
643 tests, 20 skipped, 0 unexpected passes, 4 unexpected fails, 0 flaky passes, 0 flaky fails.
UNEXPECTED FAIL:
PackageTests/ProjectImport/DedupUsingConfigFromComplex/cabal.test.hs
PackageTests/ProjectImport/UniquePathDuplicates/cabal.test.hs
PackageTests/ProjectImport/DedupUsingConfigFromSimple/cabal.test.hs
PackageTests/ProjectImport/UntrimmedImport/cabal.test.hs
...


$ ./dist-newstyle-validate-ghc-9.14.1/build/x86_64-linux/ghc-9.14.1/cabal-install-3.17.0.0/x/cabal/build/cabal/cabal --full-version
cabal-install version 3.17.0.0 (commit 9417894 on fix/y-fork-project-import-warning-only, Mon Apr 6 20:49:22 2026 -0400)
compiled using version 3.17.0.0 of the Cabal library (in-tree)
with ghc 9.14.1 on linux x86_64

$ cabal install cabal-install:exe:cabal --overwrite-policy=always

$ cabal --full-version
cabal-install version 3.17.0.0
compiled using version 3.17.0.0 of the Cabal library 
with ghc 9.14.1 on linux x86_64

$ cd cabal-testsuite/PackageTests/ProjectImport/UntrimmedImport

$ cabal build --dry-run --project-file=trailing-space.project
Warning: this is a debug build of cabal-install with assertions enabled.
Warning: import has leading or trailing whitespace;
  'https://www.stackage.org/nightly-2024-12-05/cabal.config '
    imported by: trailing-space.project
Resolving dependencies...
Build profile: -w ghc-9.14.1 -O1
In order, the following would be built (use -v for more details):
 - my-0.1 (lib:my) (first run)

$ rm -rf dist-newstyle

$ ../../../../dist-newstyle-validate-ghc-9.14.1/build/x86_64-linux/ghc-9.14.1/cabal-install-3.17.0.0/x/cabal/build/cabal/cabal build --dry-run --project-file=trailing-space.project
Warning: this is a debug build of cabal-install with assertions enabled.
Warning: import has leading or trailing whitespace;
  'https://www.stackage.org/nightly-2024-12-05/cabal.config '
    imported by: trailing-space.project
Warning: import has leading or trailing whitespace;
  'https://www.stackage.org/nightly-2024-12-05/cabal.config '
    imported by: trailing-space.project
Resolving dependencies...
Build profile: -w ghc-9.14.1 -O1
In order, the following would be built (use -v for more details):
 - my-0.1 (lib:my) (first run)
 ```

Mikolaj commented:
@philderbeast says he'd appreciate re-reviews, so I marked this accordingly.

Mikolaj commented:
From what I understand, the main reason for re-review is the introduction of the Import module with types like `type ProjectConfigSkeleton = CondTree ConfVar ([(Maybe URI, ProjectConfigPath)], ProjectConfig)`, so maybe the new reviews might focus on that. It would be great to expedite the new reviews, since the PR may be used in others.

philderbeast commented:
This could help with #11567.

philderbeast commented:
The main reason for re-review is a re-implementation after the suggestion of @mpickering, https://github.com/haskell/cabal/pull/10933#discussion_r3047495237.

geekosaur commented on cabal.validate.project:
```diff
@@ -13,7 +13,7 @@ program-options
 -- The +legacy-comparison flag is to set the default project file parser to compare
 -- the result of the legacy and parsec parser and fail if they are not equal.
 package cabal-install
-  flags: +git-rev +legacy-comparison
```
Do we still need this if the parser itself is no longer doing the validation?

ulysses4ever replied:
I think it's an unrelated change, so it should be removed from here and, optionally, put into another PR.

philderbeast replied:
The `+legacy-comparison` flag results in both parsers reporting the same error. This was causing the test to fail when it was expecting one warning but saw two (the expect one and the repeated one).

philderbeast commented:
@ulysses4ever are you able to take another look at this?

ulysses4ever commented:
@philderbeast thank you for the ping, I will see what I can do. First thing that jumps at me is the size of the PR. Do you think there's any chance to split it up a bit?..

philderbeast commented:
> First thing that jumps at me is the size of the PR.

It looks bigger than it is. I moved some stuff from `cabal-install-solver` to `cabal-install` with commit 6fe4b87778afc3c8fe5a71835db8445792296360.

philderbeast commented:
> Do you think there's any chance to split it up a bit?..

This is the last of the work started with #9578 in Dec 2023, some of which was split. I'm hoping this is the last of that batch.

ulysses4ever commented on cabal-install-solver/src/Distribution/Solver/Types/ProjectConfigPath.hs:
```diff
@@ -79,43 +82,92 @@ instance Pretty ProjectConfigPath where
-                P.<> compare (length aPaths) (length bPaths)
-                P.<> compare aPaths bPaths
+                (Nothing, Nothing) ->
+                    let aPaths = splitPath a
+                        bPaths = splitPath b
+                    in
+                        compare (length as) (length bs)
+                        P.<> compare (length asPaths) (length bsPaths)
+                        P.<> compare asPaths bsPaths
+                        P.<> compare (length aPaths) (length bPaths)
+                        P.<> compare aPaths bPaths
+                        P.<> compare aImporters bImporters
```
this suggestion looks legit, @philderbeast ?

philderbeast replied:
It is and I verified this with a doctest property test:

```
-- prop> compareSegmentally a b = compareCopilotly a b
```

 In `compareCopilotly` I removed the redundant comparison.
 
 I made the suggested change.

ulysses4ever commented on cabal-install/src/Distribution/Client/ProjectConfig/Import.hs:
```diff
@@ -0,0 +1,292 @@
+import Data.Functor ((<&>))
+import Data.List ((\\))
+import qualified Data.Map as Map
+import Distribution.Client.Compat.Prelude hiding (empty, (<>))
```
well, this went sideways...

ulysses4ever commented on changelog.d/pr-9933:
```diff
@@ -0,0 +1,26 @@
+packages: cabal-install-solver cabal-install
+significance: significant
```
I wouldn't call it a significant change, but that's of course a subjective (and non-blocking) opinion.

philderbeast replied:
I added this only when I moved some things between packages.

ulysses4ever commented on cabal-testsuite/PackageTests/ProjectImport/UniquePathDuplicates/www-stackage-org/lts-21.25.config:
do we really ned to store a 3KLOC file in our repo for tests?

philderbeast replied:
I think we do, the parsing time is real and measurable, especially when an import is duplicated, as this pull request is warning against.

ulysses4ever replied:
> the parsing time is real and measurable

this is exactly the reason i think we shouldn't have such big files in regression tests: what's the point of doing otherwise exactly? our test suite only checks correctness for the most part (there are specific parts that test performance, but PackageTests, i.e., integration tests, isn't that part).

philderbeast replied:
We also have another Stackage `*.config` file:

    cabal-testsuite/PackageTests/VersionPriority/project-stackage/nightly-2023-12-07.config

philderbeast replied:
Perhaps we can have only one example, or perhaps one that is served by our test infrastructure, or faked like `withRepo`?

https://github.com/haskell/cabal/blob/07b96a2578ae173a6919c8b29016cbc652b9b036/cabal-testsuite/src/Test/Cabal/Prelude.hs#L679-L682

philderbeast replied:
> do we really ned to store a 3KLOC file in our repo for tests?

This file is 108Kb, text-based and with repetition of `==` followed by version numbers so I'd expect it to be compressed quite well when stored in Git. I'm storing these files to avoid having our tests do `http://` imports while still using a realistic import of constraints. If the test doesn't need a realistic set of packages that work together, that the solver accepts, we could look at generating a similar-sized file just for the purpose of stressing the import parsing.

philderbeast replied:
I checked with Git on how big this data file is, uncompressed and stored:

```
$ git hash-object cabal-testsuite/PackageTests/ProjectImport/UniquePathDuplicates/www-stackage-org/lts-21.25.config
04f7e085d85d4479df8a828d9faff56ce1ddc728

$ git cat-file -s 04f7e085d85d4479df8a828d9faff56ce1ddc728
108495

$ echo 04f7e085d85d4479df8a828d9faff56ce1ddc728 | git cat-file --batch-check='%(objectsize:disk)'
6010
```

philderbeast replied:
I've gutted the contents of this file and left some notes in it:

```
$ git hash-object cabal-testsuite/PackageTests/ProjectImport/UniquePathDuplicates/www-stackage-org/lts-21.25.config
4a01c3b7a1dc19164a744a7a1426bed97297199e

$ git cat-file -s 4a01c3b7a1dc19164a744a7a1426bed97297199e
2168

$ echo 4a01c3b7a1dc19164a744a7a1426bed97297199e | git cat-file --batch-check='%(objectsize:disk)'
1019
```

The new contents:

```pre

-- NOTE: This file is intentionally gutted of contents to save space in our Git
-- repository. When it had been commited in full as downloaded, on the
-- development branch before squashing, I checked with Git on how big this data
-- file is, uncompressed and stored:
--
-- $ git hash-object cabal-testsuite/PackageTests/ProjectImport/UniquePathDuplicates/www-stackage-org/lts-21.25.config
-- 04f7e085d85d4479df8a828d9faff56ce1ddc728
--
-- $ git cat-file -s 04f7e085d85d4479df8a828d9faff56ce1ddc728
--108495
--
-- $ echo 04f7e085d85d4479df8a828d9faff56ce1ddc728 | git cat-file --batch-check='%(objectsize:disk)'
-- 6010
--
-- The point of having a project whole *.config file from Stackage was to show
-- that parsing a project import has a cost. Typically, the snapshot `*.config`
-- files Stackage provides for Cabal are quite big and costly to import. We
-- shouldn't be importing them more than once.

-- Duplicate imports by way of unique import paths are checked for after
-- parsing. I'd initially proposed that all duplicate imports should trigger an
-- error but reviewers wanted a warning instead for those imported by unique
-- import paths.  Exact duplicate imports by the same path are still checked for
-- at parsing and stop parsing with an error.
--
-- If you wish to see how bad it could get, either replace this file's contents
-- with one downloaded from Stackage (you may then need to comment out its
-- with-compiler field) or import from stackage from this file. Uncomment the
-- next line to do the latter:
-- import: https://www.stackage.org/lts-21.25/cabal.config

-- With the above import uncommented, doing a dry run build takes noticable time
-- when there are 25 imports from Stackage:
--
-- $ time cabal build all --dry-run --project-file=woops-0.project --project-file-parser=parsec
-- ...
-- Warning: 25 imports of https://www.stackage.org/lts-21.25/cabal.config;
-- ...
-- Executed in 15.97 secs
--
-- $ time cabal build all --dry-run --project-file=woops-0.project --project-file-parser=legacy
-- ...
-- Warning: 25 imports of https://www.stackage.org/lts-21.25/cabal.config;
-- ...
-- Executed in 14.10 secs

constraints: base installed
```

ulysses4ever commented on cabal-install-solver/src/Distribution/Solver/Types/ProjectConfigPath.hs:
```diff
@@ -79,43 +82,92 @@ instance Pretty ProjectConfigPath where
+    normSep p =
+        if buildOS == Windows
+            then
+                Windows.joinPath $ Windows.splitDirectories
```
could save on indentation a lot if used guards...

philderbeast replied:
I've reformatted for less indent but not used guards. This code snippet was moved, lambda lifted, but I didn't change it.

I'm surprised that this package is still in the styling TODO list:

https://github.com/haskell/cabal/blob/f444d1b9334a09a33b4c340aebfced4f31631aba/Makefile#L47

philderbeast replied:
I enabled fourmolu for `cabal-install-solver` to see how it would reformat that code.

```haskell
splitPath :: FilePath -> [FilePath]
splitPath = FP.splitPath . normSep
  where
    normSep p =
      if buildOS == Windows
        then
          Windows.joinPath $
            Windows.splitDirectories
              [if Posix.isPathSeparator c then Windows.pathSeparator else c | c <- p]
        else
          Posix.joinPath $
            Posix.splitDirectories
              [if Windows.isPathSeparator c then Posix.pathSeparator else c | c <- p]
```

philderbeast replied:
There's one less indent with guards but the extra indent with the `if` is due to fourmolu putting `then` and `else` indented on new lines.

```haskell
splitPath :: FilePath -> [FilePath]
splitPath = FP.splitPath . normSep
  where
    normSep p
      | Windows <- buildOS =
          Windows.joinPath $
            Windows.splitDirectories
              [if Posix.isPathSeparator c then Windows.pathSeparator else c | c <- p]
      | otherwise =
          Posix.joinPath $
            Posix.splitDirectories
              [if Windows.isPathSeparator c then Posix.pathSeparator else c | c <- p]
```

ulysses4ever replied:
Indentation itself isn't so much of a problem imo. My point is that guards look more idiomatic haskell.

philderbeast replied:
These changes are no longer a part of this pull request and were merged in #11773.

ulysses4ever commented:
```
- type ProjectConfigSkeleton = CondTree ConfVar ([ProjectConfigPath], ProjectConfig)
+ type ProjectConfigSkeleton = CondTree ConfVar ([(Maybe URI, ProjectConfigPath)], ProjectConfig)
```

the level of nestedness is a little much here. Would it be possible to have a type synonym (at least) for `(Maybe URI, ProjectConfigPath)`? Something like `ProjectConfigPathWithProvenance` or some such.

philderbeast commented:
```diff
- (Maybe URI, ProjectConfigPath)
+ ProjectConfigPathWithProvenance
```

The suggested type synonym is longer than the tuple and I'm not able to see what it is at a glance. Once we get rid of the legacy parser, there'll be only one type signature using this 2-tuple outside of the `Distribution.Client.ProjectConfig.Import` module:

https://github.com/haskell/cabal/blob/c3c060f4b77633b227378384ff397de09909af1d/cabal-install/src/Distribution/Client/ProjectConfig/Parsec.hs#L200

philderbeast commented:
@geekosaur  and @ulysses4ever thanks for your approvals. I've resolved all open review comments and will add the `merge me` label when it has passed the check run, unless you have any further objections?

philderbeast commented:
@mergify refresh

philderbeast commented:
@mergify queue

philderbeast commented:
This pull request was dequeued because of a failing run. I reran that failing run and it succeeded. @Mikolaj, @ulysses4ever, @geekosaur is this something that mergify could do itself, try the failing job once more before dequeuing?

The job that failed and caused the dequeue as reported earlier in this thread:

https://github.com/haskell/cabal/actions/runs/26403491730/job/77734323407

The run that passed when I reran its failed jobs:

https://github.com/haskell/cabal/actions/runs/26403491730

philderbeast commented:
@mergify queue
